clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -808,6 +808,31 @@ public boolean equals(Object o) {
         public int hashCode() {
             return partitionsPerTopic != null ? partitionsPerTopic.hashCode() : 0;
         }
+
+        public boolean topicsChanged(MetadataSnapshot that) {
+
+            if (partitionsPerTopic == null) {
+                return (that.partitionsPerTopic == null);
+            } else {
+                if (partitionsPerTopic.size() != that.partitionsPerTopic.size())
+                    return true;
+
+                for (String topic: partitionsPerTopic.keySet()) {
+                    if (that.partitionsPerTopic.containsKey(topic)) {
+                        if (that.partitionsPerTopic.get(topic) != null
+                                && this.partitionsPerTopic.get(topic) != null
+                                && !that.partitionsPerTopic.get(topic).equals(this.partitionsPerTopic.get(topic))) {
+
+                            return true;
+                        }
+                    } else {
+                        return true;
+                    }
+                }
+
+                return false;
```

Comment:
Hey @guozhangwang, what do you think about changing the structure of this method as below. My comments are inserted below:

```
public boolean topicsChanged(MetadataSnapshot that) {

       if (partitionsPerTopic == null)
           return that.partitionsPerTopic == null;

      // if that.partitionsPerTopic == null && partitionsPerTopic != null
      // then we have NPE below, right? Is that possible?

       if (partitionsPerTopic.size() != that.partitionsPerTopic.size())
           return true;

       // check if Sets are equal, as size (number of entries) is equal (per above) then it must be the same 
       if (!that.partitionsPerTopic.keySet().containsAll(partitionsPerTopic.keySet())
           return false;

       for (Entry<String,Integer> e: partitionsPerTopic.entrySet()) {
             String topic = e.getKey();
             Integer partitions = e.getValue();
             Integer thatPartitions =  that.partitionsPerTopic.get(topic);
             if (partitions  != null &&
                 // THE CHECK BELOW IS NEEDED? partitions.equals(null) is false, right?
                 thatPartitions != null && 
                 !partitions.equals(thatPartitions)) {
                 return true;
             }
        }

        return false;
}
```

The logic doesn't change (I hope so! haha), but the changes make it little bit easier to read, imho. ;)